client/writer.go

```diff
@@ -0,0 +1,376 @@
+	w.encodersByShard[shard] = encoder
+	err := w.encodeUntimedWithLock(encoder, metric, metadatas)
+	w.Unlock()
+
+	return err
+}
+
+func (w *writer) Flush() error {
+	w.RLock()
+	if w.closed {
+		w.RUnlock()
+		return errInstanceWriterClosed
+	}
+	err := w.flushWithLock()
+	w.RUnlock()
+
+	if err != nil {
+		w.metrics.flushErrors.Inc(1)
+		return err
+	}
+	return nil
+}
+
+func (w *writer) Close() error {
+	w.Lock()
+	defer w.Unlock()
+
+	if w.closed {
+		return errInstanceWriterClosed
+	}
+	w.closed = true
+	if err := w.flushWithLock(); err != nil {
+		w.metrics.flushErrors.Inc(1)
+	}
+	return w.queue.Close()
+}
+
+func (w *writer) encodeUntimedWithLock(
+	encoder *lockedEncoder,
+	metric unaggregated.MetricUnion,
+	metadatas metadata.StagedMetadatas,
+) error {
+	encoder.Lock()
+
+	var (
+		sizeBefore = encoder.Len()
+		encodeErr  error
+		enqueueErr error
+	)
+	switch metric.Type {
+	case unaggregated.CounterType:
+		msg := encoding.UnaggregatedMessageUnion{
+			Type: encoding.CounterWithMetadatasType,
+			CounterWithMetadatas: unaggregated.CounterWithMetadatas{
+				Counter:         metric.Counter(),
+				StagedMetadatas: metadatas,
+			}}
+		encodeErr = encoder.EncodeMessage(msg)
+	case unaggregated.BatchTimerType:
+		// If there is no limit on the timer batch size, write the full batch.
+		if w.maxTimerBatchSize == 0 {
+			msg := encoding.UnaggregatedMessageUnion{
+				Type: encoding.BatchTimerWithMetadatasType,
+				BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+					BatchTimer:      metric.BatchTimer(),
+					StagedMetadatas: metadatas,
+				}}
+			encodeErr = encoder.EncodeMessage(msg)
+			break
+		}
+
+		// Otherwise, honor maximum timer batch size.
+		var (
+			batchTimer     = metric.BatchTimer()
+			timerValues    = batchTimer.Values
+			numTimerValues = len(timerValues)
+			start, end     int
+		)
+		for start = 0; start < numTimerValues; start = end {
+			end = start + w.maxTimerBatchSize
+			if end > numTimerValues {
+				end = numTimerValues
+			}
+			singleBatchTimer := unaggregated.BatchTimer{
+				ID:     batchTimer.ID,
+				Values: timerValues[start:end],
+			}
+			msg := encoding.UnaggregatedMessageUnion{
+				Type: encoding.BatchTimerWithMetadatasType,
+				BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+					BatchTimer:      singleBatchTimer,
+					StagedMetadatas: metadatas,
+				}}
+			encodeErr = encoder.EncodeMessage(msg)
+			if encodeErr != nil {
+				break
+			}
+
+			// If the buffer isn't big enough continue to the next iteration.
+			if sizeAfter := encoder.Len(); sizeAfter < w.flushSize {
+				continue
+			}
+
+			// Otherwise we enqueue the current buffer.
+			buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
+
+			// Unlock the encoder before we enqueue the old buffer to ensure other
+			// goroutines have an oppurtunity to encode metrics while larger timer
+			// batches are being encoded.
+			encoder.Unlock()
+
+			enqueueErr = w.enqueueBuffer(buffer)
```

Comment:
the encoder lock usage here is unconventional and kinda hard to maintain, how much time does this enqueueBuffer call take? is it really necessary to do it outside of lock?

Reply:
Sounds good, please make sure there are test cases to guard that neither lock operation could be removed.

Reply:
Yeah this was tested in `TestWriterConcurrentWriteStress` in `writer_test.go`.